[Service Bus] Update constructors to add overloads and use Azure Identity

ramya0820 opened the pull request:
For more context refer to #4262 

Verified build and local smoke test run of live tests

ramya-rao-a commented:
I'll hold back on reviewing this PR until we rebase/merge from master after #5324 gets merged

ramya0820 commented:
Updated PR with changes from #5324 - verified that build and tests are running fine on local. @ramya-rao-a @chradek

ramya0820 commented:
> In Track2, we do not want to expose `SharedKeyCredential` and `ConnectionConfig` from our client constructors. Use the `EventHubProducerClient` constructor overloads in the Event Hubs library as reference if needed.

Updated PR 
Looks like the way we construct ConnectionCOnfig internally is by manually constructing the connection string.
This seems specific to Event Hubs and unsure if the same approach can be used for SB.

ramya-rao-a commented on sdk/servicebus/service-bus/src/serviceBusClient.ts:
```diff
@@ -56,23 +57,69 @@ export class ServiceBusClient {
+      ConnectionConfig.validate(config);
+    } else {
+      if (credentialOrServiceBusClientOptions == undefined) {
+        throw new Error("Input parameter token credential must be defined.");
```
This code path will not be hit. `isTokenCredential(undefined)` returns falsy

ramya0820 replied:
! on falsy is true right? It would go here because the if check is `(!isTokenCredential)` - So technically if someone passed in `undefined` for credentials, we'd be catching it here?

ramya-rao-a replied:
! on falsy is indeed true. When `credentialOrServiceBusClientOptions` is undefined, `if (!isTokenCredential(credentialOrServiceBusClientOptions))` will return true. Therefore, we wont enter the else block here at line 112

ramya0820 replied:
Right, mistakenly assumed the utility to check for type of the options - updated the checks

ramya-rao-a commented on sdk/servicebus/service-bus/src/serviceBusClient.ts:
```diff
@@ -56,23 +57,69 @@ export class ServiceBusClient {
-   * @param {TokenCredential} [tokenCredential] - SharedKeyCredential object or your 
+   * @param host - The host name for the Service Bus namespace. This is likely to be similar to
+   * <yournamespace>.servicebus.windows.net
+   * @param credential - SharedKeyCredential object or your
```
The overload does not support SharedKeyCredential. Please update the docs

ramya0820 replied:
Actually I had noticed this in the Event Hubs docs and thought it was being included on purpose.

Is AAD supported in Track 2? I see Event Hubs has a sample that uses `DefaultAzureCredential` 
Is that something we'll be using in Service Bus as well with `@azure/identity` (And in this PR?)

ramya-rao-a replied:
SharedKeyCredential will be re-introduced for Event Hubs post Track 2 GA. There are some design  decisions that need to be finalized there.

AAD is supported in Track 2. Thats what the use of `@azure/identity` provides us.

We don't want to update the samples just yet until we are closer to releasing Track 2 previews.

ramya-rao-a commented on sdk/servicebus/service-bus/review/service-bus.api.md:
```diff
@@ -5,12 +5,14 @@
 import Long from 'long';
 import { MessagingError } from '@azure/core-amqp';
 import { RetryOptions } from '@azure/core-amqp';
+import { SharedKeyCredential } from '@azure/core-amqp';
```
Can you run the build and ensure that the above changes to the file are indeed needed anymore?

ramya0820 replied:
updated

ramya-rao-a commented on sdk/servicebus/service-bus/src/serviceBusClient.ts:
```diff
@@ -149,33 +204,6 @@ export class ServiceBusClient {
-
   //   /**
   //    * Creates a ServiceBusClient for the Service Bus Namespace represented by the given host using
   //    * the given TokenProvider.
```
The constructor overloads that support `TokenCredential`  replace the `createFromTokenProvider` and `createFromAadTokenCredentials` methods that are commented here. These commented lines can now be removed.

The [tests for these 2 methods](https://github.com/Azure/azure-sdk-for-js/blob/9883ac911b4c9cfadd9e8d06c4837eb58c007106/sdk/servicebus/service-bus/test/serviceBusClient.spec.ts#L312-L384) that were commented need to be replaced with tests that use the `EnvironmentCredential`. See [similar tests for Event Hubs](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/test/client.spec.ts#L88-L127) for reference

ramya0820 replied:
Using `@azure/identity` in tests is causing build and rollup to fail.
Specifically, seems to do with resolving references to `RequestInit` `RequestInfo` `Response` `Headers` in `core-http` which seem to have been auto inferred from `NodeJS.Globals` / `@types/node-fetch`

Just added a draft commit to this PR if anyone is available to take a look, as this will likely involve more investigation, troubleshooting
cc: @bterlson 

The errors I'm seeing are

```
tsc -p . && rollup -c 2>&1 && npm run extract-api
../../core/core-http/es/lib/fetchHttpClient.d.ts(7,72): error TS2304: Cannot find name 'RequestInit'.
../../core/core-http/es/lib/fetchHttpClient.d.ts(9,27): error TS2304: Cannot find name 'RequestInfo'.
../../core/core-http/es/lib/fetchHttpClient.d.ts(9,47): error TS2304: Cannot find name 'RequestInit'.
../../core/core-http/es/lib/fetchHttpClient.d.ts(9,69): error TS2304: Cannot find name 'Response'.
../../core/core-http/es/lib/fetchHttpClient.d.ts(11,47): error TS2304: Cannot find name 'Headers'.
../../core/core-http/es/lib/nodeFetchHttpClient.d.ts(7,18): error TS2304: Cannot find name 'RequestInfo'.
../../core/core-http/es/lib/nodeFetchHttpClient.d.ts(7,38): error TS2304: Cannot find name 'RequestInit'.
../../core/core-http/es/lib/nodeFetchHttpClient.d.ts(7,60): error TS2304: Cannot find name 'Response'.
../../core/core-http/es/lib/nodeFetchHttpClient.d.ts(8,63): error TS2304: Cannot find name 'RequestInit'.
```

ramya-rao-a replied:
After updating the version of the `@azure/identity` to `^1.0.0` in the package.json file, I am able to build without any errors:

![image](https://user-images.githubusercontent.com/16890566/72115910-79373f80-32fd-11ea-8080-844889a9eaef.png)

ramya0820 replied:
Updated PR and verified test run on local

ramya-rao-a replied:
The commented code for the static constructors in this file still needs to be removed

ramya0820 replied:
updated

ramya-rao-a commented on sdk/servicebus/service-bus/package.json:
```diff
@@ -78,6 +78,7 @@
     "@azure/core-amqp": "1.0.0-preview.3",
     "@azure/core-http": "^1.0.0",
     "@azure/ms-rest-nodeauth": "^0.9.2",
+    "@azure/identity": "1.0.0-preview.3",
```
GA version of Identity was released back in November. Please update to use `^1.0.0` here instead of `1.0.0-preview.3`

ramya0820 replied:
Was the base branch updated manually again? 
Because the one I worked with didn't surface any install step issues. And I recall checking the version of @azure/identity to use

Also, the errors were showing up after the test code was added. Did you update the version to ^1.0.0 for code in current PR or a different setup?
I will take a look at this again to use ^1.0.0 version, verify tests on local and update the PR

ramya0820 commented:
For using @azure/identity and EnvironmentCredential, looks like the environment variables to be present are       "AZURE_CLIENT_ID", "AZURE_CLIENT_SECRET", "AZURE_TENANT_ID" - in place of the AAD_ prefixed ones that we use for Track 1.

Will we have a separate CI pipeline / configuration for track 2 branch? Or should we update track 1 to use the AZURE prefixes and then have common pipeline to run the CI?
cc: @KarishmaGhiya @mikeharder @AlexGhiondea @ramya-rao-a

KarishmaGhiya commented:
> For using @azure/identity and EnvironmentCredential, looks like the environment variables to be present are "AZURE_CLIENT_ID", "AZURE_CLIENT_SECRET", "AZURE_TENANT_ID" - in place of the AAD_ prefixed ones that we use for Track 1.
> 
> Will we have a separate CI pipeline / configuration for track 2 branch? Or should we update track 1 to use the AZURE prefixes and then have common pipeline to run the CI?
> cc: @KarishmaGhiya @mikeharder @AlexGhiondea @ramya-rao-a

@ramya0820 Have you added these to the keyvault yet? If not, please do so - add them to the keyvault  - AzureSDK-TestSecrets2 and then update the yaml files with the correct env variables required and I will do the linking on devops too

ramya-rao-a commented on sdk/servicebus/service-bus/package.json:
```diff
@@ -75,9 +75,10 @@
     ]
   },
   "dependencies": {
-    "@azure/core-amqp": "1.0.0-preview.3",
```
Change to use latest of `core-amqp` is being done as part of #6861
Any reason to include this in this PR?

ramya0820 replied:
No reason, ideally that should have been merged already - branch will get updated once that's done

ramya-rao-a replied:
Feel free to use the `allowedAlternativeVersions` section in the `common-versions.json` file to get rush to allow you to use different versions of `@azure/core-amqp` in Service Bus vs Event Hubs.

This PR is for the Azure identity work and it would help if the changes here are restricted to using the `@azure/identity` library in the constructors.

If #6861 gets merged before this PR, we can then
- rebase your branch from `feature/service-bus-track-2`
- revert the change to `common-versions.json` file

ramya0820 replied:
I looked over the diffs - the duplicate changes related to core-amqp are very few and can be ignored - it's this one line change updating version, and the type casting ones in files `messageReceiver.ts`, `messageSender.ts` and `batchingReceiver.ts`. (You can ignore the files altogether if they already appear in #6861)

All other 24-25 files are relevant and can be reviewed. The build succeeds as well.

ramya-rao-a commented on sdk/servicebus/service-bus/package.json:
```diff
@@ -75,9 +75,10 @@
-    "@azure/core-amqp": "1.0.0-preview.3",
+    "@azure/core-amqp": "^1.0.0",
     "@azure/core-http": "^1.0.0",
     "@azure/ms-rest-nodeauth": "^0.9.2",
```
`@azure/ms-rest-nodeauth` should no longer be needed now that we are using `@azure/identity`

ramya0820 replied:
It's still being used in the tests - moved it to dev dependency

ramya-rao-a replied:
After the removal of the aadutils file, I believe `@azure/ms-rest-nodeauth` is not used in tests either.

ramya-rao-a commented on sdk/servicebus/service-bus/src/serviceBusClient.ts:
```diff
@@ -56,23 +57,87 @@ export class ServiceBusClient {
+    let isConnectionStringConstructorInvoked = true;
+    if (hostOrConnectionString.indexOf("=") < 0) {
+      isConnectionStringConstructorInvoked = false;
+    }
```
The above checks are not needed. `ConnectionConfig.create()` ensures that `String()` is called on the given connection string. If the connection string thus stringified doesnt meet the requirements of a connection string, then appropriate errors are thrown from `ConnectionConfig.create()`.

I would suggest the below for handling the overloads

```javascript
if (!isTokenCredential(credentialOrServiceBusClientOptions)) {
    // assume that first overload is used & create the ConnectionConfig
} else {
    // assume that second overload is used & create the ConnectionConfig
}
// create the connection context
```

ramya0820 replied:
The reason for converting to String is because if it's a hostname then we use a .endsWith() check in here

ramya-rao-a replied:
You should still be able to use `hostOrConnectionString.endsWith()`, TS should not complain as the type is already set to `string` in the constructor signature.

If you want to avoid the case of runtime error for when a bad type is passed, then you can use `hostOrConnectionString = String(hostOrConnectionString)`. Since then endsWith() check is needed only in the case of TokenCredential, you can move this inside the if/else block as well.

ramya0820 replied:
Any reason why we want to convert it to type `String` and not use `.toString()`? If i remember correctly, I think @bterlson had once discussed, advised about not doing so and using `string` where possible.
And yes, the conversion was to help with both compile and run time scenarios.

ramya-rao-a replied:
My intention here is to simplify the logic to handle the constructor overloads. 
Currently, we are adding multiple lines of code to ensure that `hostOrConnectionString` is indeed a string and that it has `=` in it. We can avoid the need for this and the future tax of maintaining them as well.

For example, when the constructor overload is using connection string, we shouldnt need to do any checks on `hostOrConnectionString` as that work is offloaded to the downstream `ConnectionConfig.create()`.

When the constructor overload for credential is being used, us using String() will ensure that if user had provided an object with .toString() defined, then it gets called and used correctly. If the passed object does not have .toString() on it, then the connection to the service would fail eventually with the service providing an error message hinting that the provided host name is not valid.

ramya0820 replied:
I think there's a trade off with quality of expected user experience here possibly - for instance, if a user is passing in the hostname correctly and receives a connection string malformed error, wouldn't that be confusing / inappropriate?

ramya0820 replied:
For above, let's say the inputs are passed along with a malformed / incorrect / unsupported type of credentials.
Current changes address these cases. By offloading the check to service or core-amqp to expect a connection string - would make it not possible to do so.

ramya0820 replied:
Updated to use a different signature for the credential based constructor - should likely help with both simplification/maintenance and user ex concerns.

chradek replied:
I'm just catching up here and I might have missed some things as changes have been made.
1st, I remarked on this in the API file but I do think we want to keep `connectionString` and `host` as the first parameter in their respective constructors to be consistent with our other SDKs/guidelines.

I think the original comment was about the checks being done on `hostOrConnectionString`.
```
if (
      hostOrConnectionString == undefined ||
      hostOrConnectionString.toString == undefined ||
      hostOrConnectionString.toString().trim() == ""
    ) {
      throw new Error(
        "Input parameter of host or connection string must be defined and coercible to string."
      );
    }
```

I think @ramya-rao-a 's had 2 points:
1. We can ensure `hostOrConnectionString` is a string by wrapping it in `String`:  `String(hostOrConnectionString)`
Note that this is different from `new String(hostOrConnectionString)`, which is not what you want (and what I think @bterlson would warn to stay away from).
2. There are other places where `hostOrConnectionString` is validated so this check isn't needed.

The `EventHubClient` constructor shows both of these aspects off (only linking to one line here but look at the rest of the function): https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/src/impl/eventHubClient.ts#L339

Event Hubs is a bit more complicated since it has 3 overloads instead of 2, so service bus would be even simpler.

I think the Event Hubs constructor would throw helpful errors for the cases that were being checked before, but if there's something being missed let me know!

ramya-rao-a commented:
> Will we have a separate CI pipeline / configuration for track 2 branch? Or should we update track 1 to use the AZURE prefixes and then have common pipeline to run the CI?

We will re-use the same test pipeline for both track 1 and track 2

> Have you added these to the keyvault yet? If not, please do so - add them to the keyvault - AzureSDK-TestSecrets2 and then update the yaml files with the correct env variables required and I will do the linking on devops too

The secrets already exist. I believe all we need is to update the yaml files to add new entries for these env variables and point them to the same secrets as the older env variables

ramya0820 commented:
> The secrets already exist. I believe all we need is to update the yaml files to add new entries for these env variables and point them to the same secrets as the older env variables

Agree, updated

ramya-rao-a commented:
There was a [test for custom token provider](https://github.com/Azure/azure-sdk-for-js/blob/22c5103cf5f7e8333dc5ddbbaccacc75b63027eb/sdk/servicebus/service-bus/test/streamingReceiver.spec.ts#L863) which was commented in the track 2 branch because the `createFromTokenProvider` was commented in this branch.

Now that we have the constructor overload that can take a custom token provider (i.e a custom `TokenCredential`), we should re-enable the above test as well.

chradek commented on sdk/servicebus/service-bus/review/service-bus.api.md:
```diff
@@ -150,8 +150,9 @@ export class Sender {
 // @public
 export class ServiceBusClient {
+    constructor(connectionString: string, options?: ServiceBusClientOptions);
+    constructor(credential: TokenCredential, host: string, options?: ServiceBusClientOptions);
```
I think the ordering of parameters here should be switched for consistency with event hubs.
```typescript
constructor(host: string, credential: TokenCredential, options?: ServiceBusClientOptions);
```
This also matches the sample in the typescript design [guidelines](https://azure.github.io/azure-sdk/typescript_design.html#ts-apisurface-serviceclient).

ramya0820 replied:
Oh, could you point to which guideline is being considered for this?

Because I found the following that seems similar / applicable for `hostOrConnectionString` usage

```
// bad example
class ExampleClient {
  constructor (connectionString: string, options: ExampleClientOptions);
  constructor (url: string, options: ExampleClientOptions);
  constructor (urlOrCS: string, options: ExampleClientOptions) {
    // have to dig into the first parameter to see whether its
    // a url or a connection string. Not ideal.
  }
}
```

chradek replied:
Link was hidden in the word 'guidelines' :)
https://azure.github.io/azure-sdk/typescript_design.html#ts-apisurface-serviceclient

See the Client API example.
>  // client constructors have overloads for handling different
>  // authentication schemes.
>  constructor(connectionString: string, options?: ServiceClientOptions);
>  constructor(host: string, credential: TokenCredential, options?: ServiceClientOptions);

chradek replied:
The reason the example you posted is considered bad is because the type signatures for both constructors are exactly the same (string, options). In our case though we have 2 unique signatures (string, options) and (string, TokenCredential, options)

ramya0820 replied:
Unable to reply to other comment for some reason :? 
Yes, EventHubs impl seems to have the different concerns addressed. The type of error being thrown is useful to note as well (TypeError as opposed to simple Error)
I just have the guideline aspect clarification pending - ok to be consistent with EventHubs if that's what we want to do
Thoughts? @ramya-rao-a @bterlson 

Personally, I think the example called out on the guidelines page resonates with my initial thoughts on switching the parameters - also because credential based constructor sounded more right too, than host based. (Going by the first parameter to pass in).

ramya0820 replied:
Ok, the other observation I had was that options parameter being optional - seemed out of place to rely on type of to decide which constructor to use at first. So without that it can come down to two parameter constructor :?

Like, JS in general isn't strict on parameter numbers to be passed in.
So I think the one usecase that I had brought up earlier and seems to not be handled by EventHubs is if user invokes the host and credential based constructor as
(hostName, undefined, options) - then it looks like the error user would get would be about having to pass in a correctly formatted connection string.

ramya0820 replied:
To summarize the thread here - the user experience for when credential is passed in as `undefined` is the one that seems unaddressed.

Let's take a call on if we want to address that as a bug and fix it here or not - and if we want to still go ahead with mirroring EventHubs constructors for most part. 

cc @ramya-rao-a @chradek 
@bterlson @AlexGhiondea

chradek replied:
The situation where a user passed undefined for the credential but still passes the options after it seems very unlikely. I think you could make the argument that if they're explicitly trying to not pass in a TokenCredential, then they would want a connection string (so the error seems relevant). But, I don't feel too strongly since they would be going against the types and docs in this case (I'm glad we give some error).

With regards to the parameter order for the constructors, my vote is for consistency. Since event-hubs is already GA, changing the order there would need a major version bump. I see one of our goals as providing a consistent experience across the packages we own (when possible) to lower the learning curve for users going from one package to another.

ramya0820 replied:
Fair point, updating parameters ordering to keep up with consistency as if we decide to change, we can do it together

Agree that explicitly passing in undefined would map more to connection string based constructor.
But if say a user is instantiating the credentials programmatically, they could run into an `undefined` object. 
Event Hubs seems a bit different in that it accepts event hub name in addition to connection string in Event Hubs SDK. So it seems a little more relevant to Event Hubs as the err msg can be confusing to troubleshoot. (Also I wonder how long it might be before we bump version, maybe that would happen at same time as when Track 2 SB would go GA, but I digress :) )

And.. if we are of pinion that this a not so ideal state afterall, then we could have not so ideal string processing to guard against the missing/undefined token credentials case in Service Bus possibly. Let me know. For now haven't included those checks since we didn't want those. @ramya-rao-a Ok if we choose to handle it all together later as well or file a new issue? 
Because it seems like we might be regressing from Track 1 if we are no longer able to surface the error we used to be able to - ' "'credentials' is a required parameter and must be an instance of ... '

ramya0820 commented:
Had a question on about which `TokenCredential` to use in place of the SasTokenProvider from `amqp-common` that was being used previously. https://github.com/Azure/azure-sdk-for-js/pull/6770/files#r366494233 @ramya-rao-a

ramya-rao-a commented on sdk/servicebus/service-bus/test/utils/aadUtils.ts:
```diff
@@ -1,21 +1,10 @@
+  return new EnvironmentCredential();
 }
+
+export { EnvironmentCredential } from "@azure/identity";
```
Why do we need to re-export the `EnvironmentCredential` this way? Can the test files that need it not import it directly from `@azure/identity`?

ramya0820 replied:
It's because EnvironmentCredential and @azure/identity don't seem browser compatible. So needed to do this to import in the tests per browserification guidelines.

Related question/discussion at https://github.com/Azure/azure-sdk-for-js/pull/5436#issuecomment-574306134
Could you clarify on that?

ramya0820 replied:
Based on testing previously, this was added to address rollup and build issues - but with latest updates looks like the problem doesn't appear anymore - updated to remove this and aadUtils.ts files altogether

ramya-rao-a commented on sdk/servicebus/service-bus/test/utils/aadUtils.ts:
```diff
@@ -1,21 +1,10 @@
-    }
-  );
-  return tokenCreds;
+  return new EnvironmentCredential();
```
I don't see the `getTokenCredentialsFromAAD()` helper method being used anywhere anymore. Can we remove it?

ramya0820 replied:
Right, updated it

ramya-rao-a commented on sdk/servicebus/service-bus/test/serviceBusClient.spec.ts:
```diff
@@ -311,79 +301,101 @@ describe("Errors with non existing Queue/Topic/Subscription", async function():
-//   it("Coerces input to string for host in createFromAadTokenCredentials", async function(): Promise<
-//     void
-//   > {
-//     const tokenCreds = await getTokenCredentialsFromAAD();
```
Now that we don't use `getTokenCredentialsFromAAD()` anymore in this file, please remove the commented imports for the same.

ramya-rao-a commented on sdk/servicebus/service-bus/test/serviceBusClient.spec.ts:
```diff
@@ -311,79 +299,101 @@ describe("Errors with non existing Queue/Topic/Subscription", async function():
+    should.equal(errorWasThrown, true, "Error thrown flag must be true");
+  });
+
+  if (isNode) {
```
What is the reason for the `isNode()` check here?

ramya0820 replied:
That's because EnvironmentCredential is not supported in browser, so only a select subset of tests can be run in browser mode.

Carrying forward Q I had from other somewhat related comment - is `SharedKeyCredential` not going to be used and be removed from `core-amqp`? If not, was it supposed to work in browser?

ramya-rao-a replied:
Regarding the question on `SharedKeyCredential`, see https://github.com/Azure/azure-sdk-for-js/pull/6770#discussion_r366530508

Looks like among all the credentials from azure/identity, only the interactive credential works in browser which of course we can't use in CI.

There is ongoing work to change that. For now, I believe all libraries using azure/identity are falling under this case where the credentials from the azure/identity library are not testable in browser in CI.

ramya0820 commented:
/azp run js - servicebus - tests

ramya-rao-a commented on sdk/servicebus/service-bus/src/core/batchingReceiver.ts:
```diff
@@ -92,7 +92,7 @@ export class BatchingReceiver extends MessageReceiver {
         let error: Error | MessagingError;
         if (sessionError) {
-          error = translate(sessionError);
+          error = translate(sessionError) as MessagingError;
```
This casting is not required. Same for a few lines below.

ramya-rao-a replied:
This casting is not required. Same for a few lines below.

ramya-rao-a commented on sdk/servicebus/service-bus/test/README.md:
```diff
@@ -2,7 +2,9 @@
 
 - Run `npm i` to install all the dependencies of this project. This is a one time task.
 - The tests expect a series of queues, topics and subscriptions to exist in a single Service Bus instance.
```
After merging the `feature/service-bus-atom` branch, all tests now create the entities they need. So the above statement is no longer true.

```suggestion
- The tests will create the required queue, topic and subscription in the Service Bus namespace you provide using environment variables.
```

ramya-rao-a commented on sdk/servicebus/service-bus/test/README.md:
```diff
@@ -2,7 +2,9 @@
+
+  For browser environment, the connection string for the service bus should be in the environment variable `SERVICEBUS_CONNECTION_STRING_BROWSER`.
 
   Note that the tests will empty the messages in these entities to get a clean start before running each test.
```
After merging the `feature/service-bus-atom` branch, all tests now create the entities they need. So the above statement is no longer true. So this note and the table following it can be removed.

Similarly the sample content for the .env file should be cleaned up as well

ramya-rao-a commented on sdk/servicebus/service-bus/package.json:
```diff
@@ -93,6 +92,7 @@
   },
   "devDependencies": {
     "@azure/eslint-plugin-azure-sdk": "^2.0.1",
+    "@azure/ms-rest-nodeauth": "^0.9.2",
```
Now that aadutils.ts file is deleted, we don't need `@azure/ms-rest-nodeauth` in tests either right?

ramya0820 replied:
agree, updated

ramya-rao-a commented:
On second thoughts, the changes done to the readme file for test in this PR other than the ones related to changing AAD-* to AZURE-*, should be done in the master branch.

ramya0820 commented:
Agree on README related changes to be on `master`. Have opened https://github.com/Azure/azure-sdk-for-js/pull/6974 and addressed the comments and more changes on there.
